lib/middleware/authentication/authentication.test.js

```diff
@@ -0,0 +1,147 @@
+/* eslint-env jest */
+// NPM dependencies
+jest.mock('basic-auth')
+jest.mock('express/lib/response')
+const basicAuth = require('basic-auth')
+
+// Local dependencies
+let authentication = require('./authentication.js')
+
+// Local variables
+const userDetails = {
+  name: 'secret-username',
+  pass: 'secure-password'
+}
+const next = jest.fn()
+
+const res = require('express/lib/response')
+
+// Mock console.log so we can check any output
+console.error = jest.fn()
+
+describe('authentication', () => {
+  it('should be a function', () => {
+    expect(authentication).toBeInstanceOf(Function)
+  })
+
+  describe('how it runs in prod', () => {
+    beforeAll(() => {
+      process.env.NODE_ENV = 'production'
+      process.env.USE_AUTH = 'true'
+    })
+
+    describe('server with no username/password set', () => {
+      beforeAll(() => {
+        delete process.env.USERNAME
+        delete process.env.PASSWORD
+      })
+
+      beforeEach(() => {
+        console.error.mockClear()
+        authentication({}, res, next)
+      })
+
+      it('should return a console error', () => {
+        let consoleErrorMessage = console.error.mock.calls[0][0]
+        expect(consoleErrorMessage).toBe('Username or password is not set.')
+      })
+
+      it('should send marked up error message', () => {
+        let errorDisplayedToUser = res.send.mock.calls[0][0]
+        expect(errorDisplayedToUser).toBe('<h1>Error:</h1><p>Username or password not set. <a href="https://govuk-prototype-kit.herokuapp.com/docs/publishing-on-heroku#6-set-a-username-and-password">See guidance for setting these</a>.</p>')
```

Comment:
would it make sense to put this string in its own file, and be referenced from the middleware and the test? Or is that not a good test?

Reply:
I understand your point about having a single point of reference for this string but I think it would be a lot of extra machinery to have it in its own file at this stage. I think the test is fine as it is because if someone removes this line or changes the copy this test will break. As this bit of code hasn't been tested in any other places I think its fine if it breaks because someone decided to change the code. They only have one test to fix.

Reply:
cool, sounds fine to me